File: src/lib/relays/request-budget/apply.ts
import type { NostrFilter } from '../../protocol';
import {
  defaultReadPageMaxEvents,
  type ReadPageOptions,
} from '../subscription-manager-types';
import type { BudgetedFilters, RequestBudget } from './types';

export function applyBudgetToFilters(
  filters: readonly NostrFilter[],
  budget: RequestBudget,
): BudgetedFilters {
  const filterLimit = budget.filterLimit;
  if (!filterLimit) return { filters, warnings: budget.warnings };
  return {
    filters: filters.map((filter) => ({
      ...filter,
      limit: Math.min(filter.limit ?? filterLimit, filterLimit),
    })),
    warnings: budget.warnings,
  };
}

export function mergeBudgetsForRead(budgets: readonly RequestBudget[]): {
  readonly maxEvents: number;
  readonly timeoutMs: number;
} {
  return {
    maxEvents: Math.min(
      defaultReadPageMaxEvents,
      budgets.reduce((sum, budget) => sum + budget.maxEvents, 0),
    ),
    timeoutMs: Math.max(...budgets.map((budget) => budget.timeoutMs), 5000),
  };
}

export function effectiveReadOptions(
  budgets: readonly RequestBudget[],
  callerOptions: ReadPageOptions = {},
): ReadPageOptions {
  const merged = mergeBudgetsForRead(budgets);
  return {
    ...callerOptions,
    timeoutMs: callerOptions.timeoutMs ?? merged.timeoutMs,
    maxEvents: Math.min(
      callerOptions.maxEvents ?? merged.maxEvents,
      merged.maxEvents,
    ),
  };
}

File: crates/lkjstr-relays/src/request_budget/apply.rs
#![doc = "Request-budget application helpers."]

use lkjstr_protocol::NostrFilter;

use super::{BudgetedFilters, MergedReadBudget, RequestBudget, default_read_page_max_events};

#[must_use]
pub fn apply_budget_to_filters(filters: &[NostrFilter], budget: &RequestBudget) -> BudgetedFilters {
    let Some(filter_limit) = budget.filter_limit else {
        return BudgetedFilters {
            filters: filters.to_vec(),
            warnings: budget.warnings.clone(),
        };
    };
    BudgetedFilters {
        filters: filters
            .iter()
            .map(|filter| budget_filter(filter, filter_limit))
            .collect(),
        warnings: budget.warnings.clone(),
    }
}

#[must_use]
pub fn merge_budgets_for_read(budgets: &[RequestBudget]) -> MergedReadBudget {
    let max_events = budgets
        .iter()
        .fold(0_u64, |sum, budget| sum.saturating_add(budget.max_events))
        .min(default_read_page_max_events());
    let timeout_ms = budgets
        .iter()
        .map(|budget| budget.timeout_ms)
        .max()
        .map_or(5000, |timeout| timeout.max(5000));
    MergedReadBudget {
        max_events,
        timeout_ms,
    }
}

fn budget_filter(filter: &NostrFilter, filter_limit: u64) -> NostrFilter {
    let mut next = filter.clone();
    next.limit = Some(
        filter
            .limit
            .map_or(filter_limit, |limit| limit.min(filter_limit)),
    );
    next
}
